### ISSUE TITLE
Implement a pull parser

### ISSUE THREAD

**Issue by hadley** [hadley/xml2#10]
For large xml files, or when you just want to extract a little bit of data without loading the entire file into memory, e.g. http://lxml.de/parsing.html#incremental-event-parsing

If you need this, please :+1: this issue with a brief use case

**Comment by JakeRuss** [comment]
Google and USPTO provide [patents data](http://www.google.com/googlebooks/uspto-patents-grants-text.html) as concatenated xml files (~0.5 GB/ per file). I attempted to read in one of these, 
```
doc <- read_xml("~/Downloads/ipg130129.xml")
```
And was met with the following message,
"Error: XML declaration allowed only at the start of the document [64]"

After looking around the web I see that this means I need to split the concatenated xml file into separate smaller xml files and then parse the each of those individually.

Would your pull parser help me here? Or do I need to look into some other tool?

**Comment by hadley** [comment]
@JakeRuss no, that's different problem.

**Comment by cboettig** [comment]
:+1: This would be awesome (if I've understood it correctly).  I often work with some large dump of XML files, like this: https://github.com/TreeBASE/supertreebase/tree/master/data/treebase 

The standard use case is that I want to filter the collection to isolate only those files that have certain attributes, e.g. identifying the documents that include my species of interest as one of the taxa (e.g. matches the `<otu label="species name">` value).  

Loading all the documents into memory first is often prohibitive in terms of memory, forcing the user to parse, search, and then remove each object, and at least with `XML` this can all be very slow.

**Comment by jimhester** [comment]
libxml2 documentation of the SAX interface is http://www.jamesh.id.au/articles/libxml-sax/libxml-sax.html

The relevant Cython code from lxml is https://github.com/lxml/lxml/blob/572e10843774a5d6300125d89bdc423d53c92971/src/lxml/saxparser.pxi

Implementing this is clearly non-trivial and has entirely different semantics than we are currently using (SAX callbacks vs tree / DOM based).

This feature is a long way away and may be better implemented in a separate package entirely.

**Comment by hadley** [comment]
Let's close for now.

**Comment by randomgambit** [comment]
hey mister @jimhester , any news on this? ultra large XML files are becoming the norm today... SAX looks like a necessity even tho I dont like saxophones (haha)

Thanks!

**Comment by jimhester** [comment]
> This feature is a long way away and may be better implemented in a separate package entirely.

**Comment by randomgambit** [comment]
dude its been more than a year!!! :) Im kidding. thanks anyway for the headsup

**Comment by wkumler** [comment]
Upvoting this issue! I know it's been a long while but a new instrument we've got in the lab produces 1-10 GB XML files that need to be parsed and while I can currently use `read_xml` without any major problems, trying to find specific nodes with `find_xml_all` or `find_xml_first` rapidly consumes all available memory and errors out. Currently switching to the XML library for this but I've found the syntax is a lot messier there.
